Make WebCore::Frame a main-thread-only object

achristensen07 opened the pull request:
#### e3627f43f3e6863a69c9a53bb3a562085358b93e
<pre>
Make WebCore::Frame a main-thread-only object
<a href="https://bugs.webkit.org/show_bug.cgi?id=293141">https://bugs.webkit.org/show_bug.cgi?id=293141</a>
<a href="https://rdar.apple.com/104512003">rdar://104512003</a>

Reviewed by Geoffrey Garen and David Kilzer.

ThreadSafeRefCounted and CanMakeWeakPtr is a recipe for thread safety issues.
The common solution is to remove CanMakeWeakPtr and add ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr
and replace WeakPtr&lt;T&gt; with ThreadSafeWeakPtr&lt;T&gt;.  With WebCore::Frame, however,
that would require a lot of changes and I think it would be heading in the wrong
direction.  Instead, we remove ThreadSafeRefCounted and make it RefCounted instead.
It was originally made ThreadSafeRefCounted in 160925@main and I don&apos;t think it needs to be.

This has several benefits:
1. If it&apos;s ever used off the main thread, we will see debug assertions that will help us fix
the issue.  It shouldn&apos;t ever be used off the main thread.
2. It makes ref/deref no longer need to be an atomic operation, which means it&apos;s a little faster.
3. The compiler can now elide ref/deref pairs it can prove aren&apos;t necessary, which means it&apos;s
even faster.  The only operation faster than 1 operation is 0 operations.

* Source/WebCore/page/Frame.h:

Canonical link: <a href="https://commits.webkit.org/295131@main">https://commits.webkit.org/295131@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7074306d7680fa72c482dfb0eda2ea80dac92580

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104126 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23830 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14150 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109322 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54794 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106166 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24198 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32374 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79102 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107132 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18798 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93937 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59429 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18606 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11979 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54154 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88321 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12034 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111708 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31282 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23069 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88113 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31646 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90125 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87773 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22352 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32660 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10411 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25746 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31211 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36524 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31005 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34341 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32565 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer reviewed (approved):
r=me, but you should probably have someone else review as well.

Can we change Debug asserts to Release asserts, or do they have perf impact?

geoffreygaren commented on Source/WebCore/page/Frame.h:
```diff
@@ -58,7 +58,7 @@ enum class ScrollbarMode : uint8_t;
 using SandboxFlags = OptionSet<SandboxFlag>;
 
-class Frame : public ThreadSafeRefCounted<Frame, WTF::DestructionThread::Main>, public CanMakeWeakPtr<Frame> {
+class Frame : public RefCounted<Frame>, public CanMakeWeakPtr<Frame> {
```
This can be RefCountedAndCanMakeWeakPtr.

geoffreygaren reviewed (approved):
r=me

probably need to double check the ews failures

achristensen07 committed:
Make WebCore::Frame a main-thread-only object
https://bugs.webkit.org/show_bug.cgi?id=293141
rdar://104512003

Reviewed by Geoffrey Garen and David Kilzer.

ThreadSafeRefCounted and CanMakeWeakPtr is a recipe for thread safety issues.
The common solution is to remove CanMakeWeakPtr and add ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr
and replace WeakPtr<T> with ThreadSafeWeakPtr<T>.  With WebCore::Frame, however,
that would require a lot of changes and I think it would be heading in the wrong
direction.  Instead, we remove ThreadSafeRefCounted and make it RefCounted instead.
It was originally made ThreadSafeRefCounted in 160925@main and I don't think it needs to be.

This has several benefits:
1. If it's ever used off the main thread, we will see debug assertions that will help us fix
the issue.  It shouldn't ever be used off the main thread.
2. It makes ref/deref no longer need to be an atomic operation, which means it's a little faster.
3. The compiler can now elide ref/deref pairs it can prove aren't necessary, which means it's
even faster.  The only operation faster than 1 operation is 0 operations.

* Source/WebCore/page/Frame.h:

Canonical link: https://commits.webkit.org/295131@main
